### PR TITLE
assorted GLSL code printing improvements

### PR DESCRIPTION
**Pull request by micahscopes**
#### Brief description of what is fixed or changed
I made some improvements to the GLSL printer, along with a small improvement to the generic `codeprinter` submodule.

In `sympy.printing.codeprinter`:
- Allow custom multiline "spread" assignments when printing array-like objects, e.g.:
```
    >>> x_struct_members = symbols('x.a x.b x.c x.d')
    >>> print(glsl_code(Matrix([1,2,3,4]), assign_to=x_struct_members))
    x.a = 1;
    x.b = 2;
    x.c = 3;
    x.d = 4;
```

In `sympy.printing.glsl`:
- allow passing in a `array_type` parameter, to support printing to various GLSL array types (e.g. `float[N](...)`, `int[N](...)`)
- [bugfix] ensure that when operators are printed as functions, i.e. when `3 - x` is printed as `sub(3, x)`, that strictly negative valued expressions print correctly.  For example: `-x` will now print as `sub(0, x)`
  - allow for a custom zero expression (e.g. `zero()`) to support printing expressions of abstract GLSL types.

#### Other comments
I've included tests and inline doctest examples!

#### Release Notes

* printing
  * Allow spreading assignments across multiple symbols when printing multi-member objects.
  * Support various array constructor types when printing arrays to GLSL.
  * Fixes a bug when printing negative expressions to GLSL with operators printed as functions.
  * Support a custom `0` substitution for printing expressions representing various GLSL types.

### CODE REVIEW

**Comment by micahscopes** [comment]
c.c. @bjodah @asmeurer 

I made some improvements to the GLSL code printer, along with a little "spread assignment" feature for the generic code printer.   Let me know if you have any questions!

File: sympy/printing/codeprinter.py
Comment (eric-wieser) [inline sympy/printing/codeprinter.py:97]: It strikes me as odd that the str -> symbol logic is not the same for lists and single elements. Perhaps something like:
```python
def _handle_assign_to(expr, assign_to):
    if assign_to is None:
        return sympify(expr)
    if isinstance(assign_to, (list, tuple)):
        if len(expr) != len(assign_to):
            raise ValueError(...)
        return CodeBlock([_handle_assign_to(rhs, lhs) for lhs, rhs in zip(expr, assign_to)
    if isinstance(assign_to, str):
        assign_to = ... # as before
    elif not isinstance(assign_to, Basic):
        raise ... # as before
    return Assignment(assign_to, expr)

expr = _handle_assign_to(expr, assign_to)
```
Which would _maybe_ support nested tuple targets too.

File: sympy/printing/glsl.py
Comment (eric-wieser) [inline sympy/printing/glsl.py:313]: Which part are you expecting to raise?

File: sympy/printing/glsl.py
Comment (eric-wieser) [inline sympy/printing/glsl.py:435]: Why not just make the argument `int`? Aren't the square brackets always required?

**Comment by oscarbenjamin** [comment]
It looks like there are some useful changes here but I see that there are unaddressed comments as well.

@micahscopes what's the status of this? Are you still working on it?

**Comment by micahscopes** [comment]
@oscarbenjamin I haven't made time to give this a good look since @eric-wieser left those helpful comments.  Hopefully soon!

**Comment by czgdp1807** [comment]
@micahscopes Would you like to conclude this PR? Please address the comments if so, otherwise, do not close this PR. Thanks for your contributions.

File: sympy/printing/glsl.py
Comment (micahscopes) [reply sympy/printing/glsl.py]: Hey, this was really good feedback.  I got rid of the `array_constructor` param and replaced it with a `base_type` parameter, as you've suggested.

Your insight to use a custom class is awesome... I'll have to try that soon!

File: sympy/printing/glsl.py
Comment (micahscopes) [reply sympy/printing/glsl.py]: This has been changed

File: sympy/printing/codeprinter.py
Comment (micahscopes) [reply sympy/printing/codeprinter.py]: Okay!  Nice suggestion, I implemented this in the last commit, and basic nested tuple/list assignments work just as you'd imagined :partying_face:.  It was simplest to make a test case in the `test_glsl.py` file, since `codeprinter.py` is pretty abstract and doesn't implement all the basic printing functions.  Hopefully that's alright.

**Comment by micahscopes** [comment]
> @micahscopes Would you like to conclude this PR? Please address the comments if so, otherwise, do not close this PR. Thanks for your contributions.

I finally got around to addressing all of @eric-wieser's comments, so I think this should be just about finished up.

File: sympy/printing/tests/test_glsl.py
Comment (eric-wieser) [inline sympy/printing/tests/test_glsl.py:424]: Indent is less jarring as
```suggestion
    assert gl(A, assign_to = assign_to) == textwrap.dedent('''\
        x.a = 1;
        x.b = 2;
        x.c = 3;
        x.d = 4;
        x.e = 5;
        x.f = 6;
        x.g = 7;''')
```
(needs `import textwrap` at the top if not already there)

File: sympy/printing/tests/test_glsl.py
Comment (eric-wieser) [inline sympy/printing/tests/test_glsl.py:25]: What did this output before?

File: sympy/printing/tests/test_glsl.py
Comment (eric-wieser) [inline sympy/printing/tests/test_glsl.py:411]: PEP8 says
```suggestion
    assert gl(A, array_type='AbcType', glsl_types=False) == 'AbcType[3](a, b, c)'
```

File: sympy/printing/tests/test_glsl.py
Comment (eric-wieser) [inline sympy/printing/tests/test_glsl.py:406]:
```suggestion
    assert gl(A, array_type='int') == 'int[7](1, 2, 3, 4, 5, 6, 7)'
```

File: sympy/printing/glsl.py
Comment (eric-wieser) [inline sympy/printing/glsl.py:114]: `Template` seems like massive overkill here. How about:
```suggestion
        array_constructor = lambda size: '{}[{}]'.format(array_type, size)
```

File: sympy/printing/tests/test_glsl.py
Comment (eric-wieser) [reply sympy/printing/tests/test_glsl.py:25]: Ah, I see it crashed!

File: sympy/printing/glsl.py
Comment (micahscopes) [reply sympy/printing/glsl.py:114]: good point :sweat_smile:

File: sympy/printing/tests/test_glsl.py
Comment (eric-wieser) [reply sympy/printing/tests/test_glsl.py:25]: A test without `zero=` would be good here too.

File: sympy/printing/tests/test_glsl.py
Comment (eric-wieser) [inline sympy/printing/tests/test_glsl.py:424]: The whole point of my `textwrap.dedent` suggestion was to allow you to indent these lines!

File: sympy/printing/tests/test_glsl.py
Comment (eric-wieser) [inline sympy/printing/tests/test_glsl.py:881]: If you're going to change _all_ these tests, it would make more sense to split this into two PRs:

* One that makes these stylistic changes only
* One that introduces the new GLSL features.

**Comment by micahscopes** [comment]
@eric-wieser should be good now

File: sympy/printing/tests/test_glsl.py
Comment (micahscopes) [reply sympy/printing/tests/test_glsl.py]: I've reverted those extra changes.

File: sympy/printing/tests/test_glsl.py
Comment (eric-wieser) [inline sympy/printing/tests/test_glsl.py:469]:
```suggestion
      h = 8;'''
    )
```

Comment (eric-wieser) [inline sympy/printing/tests/test_glsl.py:469]: Indents should be multiples of four space - here I see 6, 7, and 8-space indents.

**Review by eric-wieser — APPROVED** [review]
As discussed offline, this looks good - would be great to see a test for the exceptions, but if you don't get around to it soon, then better to merge it anyway than do nothing.

**Comment by micahscopes** [comment]
> As discussed offline, this looks good - would be great to see a test for the exceptions, but if you don't get around to it soon, then better to merge it anyway than do nothing.

I added a test for the mismatched length exception.  I think this is quite ready to merge now.  Thanks again for the very thorough review @eric-wieser!

**Comment by micahscopes** [comment]
@czgdp1807 this should be ready to go!

**Comment by czgdp1807** [comment]
Will merge tomorrow if no objections raised and if it doesn't get merged until then.

**Comment by eric-wieser** [comment]
Travis seems unwilling to schedule a build. Is sympy out of travis credits?

**Comment by oscarbenjamin** [comment]
We are migrating to Github Actions from Travis: https://github.com/sympy/sympy/issues/20374

Travis has significantly reduced the number of workers available to open source projects at least on their current domain. The result is a backlog of jobs which varies on a daily/weekly basis but seems to be particularly high today:
https://www.traviscistatus.com/#month

**Comment by oscarbenjamin** [comment]
Closing and reopening to restart the tests

**Comment by micahscopes** [comment]
Closing and re-opening to (hopefully) kick off the Travis build again...

**Comment by micahscopes** [comment]
@czgdp1807 this is ready to go :+1:

**Comment by czgdp1807** [comment]
@eric-wieser Would you like to suggest any further changes in this PR? There are some new commits after your approval.

**Comment by micahscopes** [comment]
@czgdp1807 I think for the most part, those commits showing since the approval were the result of my force push after rebasing (to get rid of Travis configuration so that the CI would pass).

There was one substantive change since the approval, but it was one I'd discussed with @eric-wieser (an additional test).

**Pull request merged**
